cloud-controller-manager/do/lb_service_admission_handler.go

```diff
@@ -0,0 +1,171 @@
+	}
+
+	return resp
+}
+
+func (h *LBServiceAdmissionHandler) handle(ctx context.Context, req admission.Request) admission.Response {
+	var svc corev1.Service
+	err := h.decoder.Decode(req, &svc)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed to decode admission request: %w", err))
+	}
+
+	if svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return admission.Allowed("allowing service that is not a load balancer")
+	}
+
+	if svc.DeletionTimestamp != nil {
+		return admission.Allowed("allowing service that is being deleted")
+	}
+
+	lbReq, err := h.buildLoadBalancerRequest(&svc)
+	if err != nil {
+		return admission.Denied(fmt.Sprintf("failed to build DO API request: %s", err))
+	}
+
+	var resp admission.Response
+	switch {
+	case req.OldObject.Raw != nil:
+		resp = h.validateUpdate(ctx, svc, req, lbReq)
+	default:
+		resp = h.validateCreate(ctx, svc, req, lbReq)
+	}
+
+	return resp
+}
+
+func (h *LBServiceAdmissionHandler) validateUpdate(ctx context.Context, svc corev1.Service, req admission.Request, lbReq *godo.LoadBalancerRequest) admission.Response {
+	var oldSvc corev1.Service
+	if err := h.decoder.DecodeRaw(req.OldObject, &oldSvc); err != nil {
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode old object: %w", err))
+	}
+
+	// We ignore errors when building the old service's godo request because
+	// it is allowed to be wrong. In cases where it errors, it can potentially
+	// get fixed after the update.
+	oldReq, _ := h.buildLoadBalancerRequest(&oldSvc)
+	if cmp.Equal(oldReq, lbReq) {
+		return admission.Allowed("new service has insignificant changes")
+	}
+
+	lbID := oldSvc.Annotations[annDOLoadBalancerID]
+	if lbID == "" {
+		return admission.Denied(fmt.Sprintf("annotation %q is not set", annDOLoadBalancerID))
+	}
+
+	_, resp, err := h.godoClient.LoadBalancers.Update(ctx, lbID, lbReq)
+	return h.mapGodoRespToAdmissionResp(resp, err)
+}
+
+func (h *LBServiceAdmissionHandler) validateCreate(ctx context.Context, svc corev1.Service, req admission.Request, lbReq *godo.LoadBalancerRequest) admission.Response {
+	_, resp, err := h.godoClient.LoadBalancers.Create(ctx, lbReq)
+	return h.mapGodoRespToAdmissionResp(resp, err)
+}
+
+func (h *LBServiceAdmissionHandler) buildLoadBalancerRequest(svc *corev1.Service) (*godo.LoadBalancerRequest, error) {
+	lbReq, err := buildLoadBalancerRequest(svc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build base load balancer request: %s", err)
+	}
+	lbReq.ValidateOnly = true
+	lbReq.Region = h.region
+	lbReq.VPCUUID = h.vpcID
+	lbReq.Tags = []string{buildK8sTag(h.clusterID)}
+	return lbReq, nil
+}
+
+func (h *LBServiceAdmissionHandler) mapGodoRespToAdmissionResp(resp *godo.Response, err error) admission.Response {
+	switch {
+	case err == nil:
+		return admission.Allowed("valid load balancer definition")
+	case resp == nil: // This case should never happen, asserting to prevent nil pointer exceptions.
```

Comment:
This can legitimately happen when a networking-related error occurred (e.g., I/O problem). I think we want to include `err` in that case.